### PR TITLE
Add `url` secondary option to any rules

### LINKED ISSUE #7484: Add `url` secondary option to any rules
### What is the problem you're trying to solve?

Some generic rules (e.g. rule-selector-property-disallowed-list) can be used multiple times, for different reasons, but can link to only one URL. As it is, rule authors cannot point to the relevant documentation for each different use of the rule.

### What solution would you like to see?

Enabling the use of custom URLs in the config file, like with messages.

### PR DESCRIPTION
**Pull request by emmacharp**
> Which issue, if any, is this issue related to?

Closes #7484

> Is there anything in the PR that needs further explanation?

Based on @ybiquitous implementation proposal [here](https://github.com/stylelint/stylelint/issues/7484#issuecomment-2147707546). Additional changes were made to `lintSource.mjs` and `validateOptions.mjs` as to prevent undefined errors and invalid option warnings. `customUrls` property was added to `checkAgainstRule.mjs` as to pass certain tests.

### CODE REVIEW

**Comment by ybiquitous** [comment]
@emmacharp Thanks for the pull request! First of all, you can fix the system test failures by running the command:

```shell
npx jest system-tests --updateSnapshot
```

File: lib/utils/report.mjs [hunk 1]
```diff
@@ -113,6 +113,12 @@ export default function report(problem) {
 	}
 
 	const { customMessages } = result.stylelint;
+	const { customUrls } = result.stylelint;
```
Comment (ybiquitous) [inline lib/utils/report.mjs:116]: [suggest] Can you shorten the code a bit like this?

```suggestion
	const { customMessages, customUrls } = result.stylelint;
```

**Review by ybiquitous — CHANGES_REQUESTED** [review]
[suggest] We should add a new section for `url` under the `rules` section in the documentation:
https://github.com/stylelint/stylelint/blob/9a2e6540f92509628df76032262293ae57d32dda/docs/user-guide/configure.md#rules

**Comment by ybiquitous** [comment]
[ask] @emmacharp Just to be sure, how will you specifically configure your rules when this `url` option is released?
I'm just wondering if you need the function support as described on <https://github.com/stylelint/stylelint/issues/7484#issuecomment-1902608789>.

**Comment by emmacharp** [comment]
@ybiquitous, I do not see the need for function support as of now.

**Comment by emmacharp** [comment]
> ```shell
> npx jest system-tests --updateSnapshot
> ```

This command is giving me errors. For instance:

```
 SyntaxError: Cannot use import statement outside a module
   at Runtime.createScriptFromCode (node_modules/jest-runtime/build/index.js:1505:14)
```

Any idea why?

File: docs/user-guide/configure.md [hunk 2]
```diff
@@ -172,6 +172,23 @@ Experimental feature: some rules support message arguments. For example, when co
...
+```json
+{
+  "color-no-hex": [
+    true,
+    {
+      "url": "https://stylelint.io"
+    }
+  ]
+}
+```
```
Comment (ybiquitous) [inline docs/user-guide/configure.md:190]: [suggest] Thanks, the added description sounds good! Additionally, what about the following points?

- Fix a typo in "For example".
- Save a few lines in the code example.
- Use another domain other than `stylelint.io` to clarify the purpose of this option.

````suggestion
For example:

```json
{
  "color-no-hex": [true, {"url": "https://example.org/your-custom-doc"}]
}
```
````

**Comment by ybiquitous** [comment]
Hum, I've never seen this error `SyntaxError: Cannot use import statement outside a module` ... 🤔 

I might be wrong, but it seems due to environmental issues. Here's my environment (macOS latest)

```sh-session
$ node -v && npm -v
v22.2.0
10.8.1
```

And the clean installation and updating Jest snapshots work for me. 

```sh-session
$ npm ci
...
$ npx jest system-tests --updateSnapshot
...
$ git status
On branch custom_urls
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   system-tests/002/__snapshots__/fs.test.mjs.snap
	modified:   system-tests/002/__snapshots__/no-fs.test.mjs.snap
	modified:   system-tests/003/__snapshots__/fs.test.mjs.snap
	modified:   system-tests/003/__snapshots__/no-fs.test.mjs.snap
	modified:   system-tests/004/__snapshots__/fs.test.mjs.snap
	modified:   system-tests/004/__snapshots__/no-fs.test.mjs.snap
	modified:   system-tests/005/__snapshots__/commonjs.test.cjs.snap

no changes added to commit (use "git add" and/or "git commit -a")
```

**Review by ybiquitous — CHANGES_REQUESTED** [review]
[suggest] Can you add a new test file named `lib/__tests__/standalone-custom-url.mjs` to test this new feature? If you prefer another file name, you may take it.

File: docs/user-guide/configure.md (hunk 2 above)
Comment (emmacharp) [reply docs/user-guide/configure.md]: Boy, embarrassing typo there. Done.

**Comment by emmacharp** [comment]
I've been able to run jest by following the second option here:
[](https://jestjs.io/docs/ecmascript-modules)

File: system-tests/003/__snapshots__/fs.test.mjs.snap
Comment (ybiquitous) [inline system-tests/003/__snapshots__/fs.test.mjs.snap:1]: [suggest] For some reason, this file was updated, and the tests on CI failed. Maybe, reverting this file change should make them successful again?

**Comment by ybiquitous** [comment]
Oh, I forgot about `--experimental-vm-modules` 😅 

https://github.com/stylelint/stylelint/blob/7a78847e959f195cb545fb9869d0641efc1457dd/package.json#L77

So, we should run Jest like this (of course, `NODE_ENV` is also available):

```shell
npm run test --ignore-scripts -- --updateSnapshot system-tests/
```

**Comment by emmacharp** [comment]
> [suggest] Can you add a new test file named `lib/__tests__/standalone-custom-url.mjs` to test this new feature? If you prefer another file name, you may take it.

I'm not really familiar with these kinds of test but I'll give it a shot and report back!

**Comment by ybiquitous** [comment]
This test file might be helpful: 👇🏼 
https://github.com/stylelint/stylelint/blob/3bdb897f647b776973abd630114141f96990c2e6/lib/__tests__/standalone-quiet.test.mjs

If you find it too hard, feel free to ask for help. If you don't mind, I can also do that instead of you. 😃

**Comment by emmacharp** [comment]
Reporting back!

I've created a test file [here](https://github.com/stylelint/stylelint/pull/7743/commits/56efa0c792816c0b60ac00489d1ad1d60d452f7b) which seems to work as intended. Still new to Jest so I hope it really is. Hehe.

**Review by ybiquitous — APPROVED** [review]
Thank you! LGTM 👍🏼

**Comment by Mouvedia** [comment]
> @ybiquitous, I do not see the need for function support as of now.

The only use case that I can see would be to dynamically change the anchor/fragment.
This can be postponed.

**Pull request merged**